### PR TITLE
Charlesmchen/tweak messages1

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/Cells/ConversationViewCell.h [hunk 1]
```diff
@@ -69,12 +70,11 @@ NS_ASSUME_NONNULL_BEGIN
 // * Users enters another view (e.g. conversation settings view, call screen, etc.).
 @property (nonatomic) BOOL isCellVisible;
 
-// The width of the collection view.
-@property (nonatomic) int contentWidth;
+@property (nonatomic, nullable) ConversationLayoutInfo *layoutInfo;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/ConversationViewCell.h]: This PR introduces `ConversationLayoutInfo`, which gathers layout-related metadata like sizes, spacing, margins.  I suspect it will get elaborated.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSBubbleView.h [hunk 2]
```diff
@@ -8,10 +8,6 @@ NS_ASSUME_NONNULL_BEGIN
 
 extern const CGFloat kOWSMessageCellCornerRadius;
 
-extern const CGFloat kBubbleVRounding;
-extern const CGFloat kBubbleHRounding;
-extern const CGFloat kBubbleThornSideInset;
-extern const CGFloat kBubbleThornVInset;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSBubbleView.h]: Tweaked the thorns.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m [hunk 3]
```diff
@@ -38,16 +41,17 @@ - (void)commontInit
 {
     OWSAssert(!self.titleLabel);
 
-    //    [self setTranslatesAutoresizingMaskIntoConstraints:NO];
+    self.preservesSuperviewLayoutMargins = NO;
+    self.contentView.preservesSuperviewLayoutMargins = NO;
+    self.layoutMargins = UIEdgeInsetsZero;
+    self.contentView.layoutMargins = UIEdgeInsetsZero;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: Converted this cell to use auto layout.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m [hunk 4]
```diff
@@ -85,30 +106,55 @@ + (NSString *)cellReuseIdentifier
...
-{
-    return UIFont.ows_dynamicTypeBodyFont.ows_mediumWeight;
-}
-
-- (UIFont *)buttonFont
-{
-    return UIFont.ows_dynamicTypeBodyFont;
-}
+    CGFloat buttonWidth = 0.f;
+    if (interaction.hasAddToContactsOffer) {
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: In the next PR, I'm going to just manipulate the button's `hidden` property.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m [hunk 5]
```diff
@@ -85,30 +106,55 @@ + (NSString *)cellReuseIdentifier
...
+    CGFloat buttonHeight = self.buttonHeight;
+    [NSLayoutConstraint deactivateConstraints:self.layoutConstraints];
+    self.layoutConstraints = @[
+        [self.addToContactsButton autoSetDimension:ALDimensionWidth toSize:buttonWidth],
+        [self.addToProfileWhitelistButton autoSetDimension:ALDimensionWidth toSize:buttonWidth],
+        [self.blockButton autoSetDimension:ALDimensionWidth toSize:buttonWidth],
+
+        [self.addToContactsButton autoSetDimension:ALDimensionHeight toSize:buttonHeight],
+        [self.addToProfileWhitelistButton autoSetDimension:ALDimensionHeight toSize:buttonHeight],
+        [self.blockButton autoSetDimension:ALDimensionHeight toSize:buttonHeight],
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: I might rework this to use a second stack view, which would hopefully let us skip this sizing.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 6]
```diff
@@ -855,41 +850,46 @@ - (void)showAttachmentErrorViewWithMediaView:(UIView *)mediaView
...
+- (CGSize)bodyTextSize:(BOOL)includeMargins
 {
+    OWSAssert(self.layoutInfo);
+    OWSAssert(self.layoutInfo.maxMessageWidth > 0);
+
     if (!self.hasBodyText) {
         return CGSizeZero;
     }
 
-    BOOL isRTL = self.isRTL;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: I'm trying to slowly converge on using `CurrentAppContext.isRTL`.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 7]
```diff
@@ -926,28 +926,32 @@ - (CGSize)bodyMediaSizeForContentWidth:(int)contentWidth
...
+            result = CGSizeMake(
                 maxMessageWidth, [OWSContactShareView bubbleHeightForContactShare:self.viewItem.contactShare]);
+            break;
     }
-}
 
-- (int)maxMessageWidthForContentWidth:(int)contentWidth
-{
-    return (int)floor(contentWidth * 0.8f);
+    return CGSizeCeil(result);
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: I'm trying to move towards using `CGSizeCeil()` to ensure point-align our layouts where possible.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m [hunk 8]
```diff
@@ -22,6 +23,8 @@ @interface OWSSystemMessageCell ()
 
 @property (nonatomic) UIImageView *imageView;
 @property (nonatomic) UILabel *titleLabel;
+@property (nonatomic) UIStackView *stackView;
+@property (nonatomic) NSArray<NSLayoutConstraint *> *layoutConstraints;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m]: Converted this to use auto layout and stack view.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSUnreadIndicatorCell.m [hunk 9]
```diff
@@ -15,12 +16,9 @@ @interface OWSUnreadIndicatorCell ()
...
 @property (nonatomic, nullable) TSUnreadIndicatorInteraction *interaction;
 
-@property (nonatomic) UIView *bannerView;
-@property (nonatomic) UIView *bannerTopHighlightView;
-@property (nonatomic) UIView *bannerBottomHighlightView1;
-@property (nonatomic) UIView *bannerBottomHighlightView2;
 @property (nonatomic) UILabel *titleLabel;
-@property (nonatomic) UILabel *subtitleLabel;
+@property (nonatomic) UIView *strokeView;
+@property (nonatomic) NSArray<NSLayoutConstraint *> *layoutConstraints;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSUnreadIndicatorCell.m]: Started to tweak this indicator.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 10]
```diff
@@ -252,6 +260,14 @@ - (nullable ConversationViewCell *)measurementCell
     return measurementCell;
 }
 
+- (CGFloat)vSpacingWithLastLayoutItem:(id<ConversationViewLayoutItem>)lastLayoutItem
+{
+    OWSAssert(lastLayoutItem);
+
+    // TODO:
+    return 0.f;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: I'm going to tweak intra-cell spacing here.

File: Signal/src/ViewControllers/ConversationView/ConversationLayoutInfo.swift [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+    @objc public let contentMarginTop: CGFloat = 10
+    @objc public let contentMarginBottom: CGFloat = 10
+
+    @objc public var gutterLeading: CGFloat = 0
+    @objc public var gutterTrailing: CGFloat = 0
+    // These are the gutters used by "full width" views
+    // like "date headers" and "unread indicator".
+    @objc public var fullWidthGutterLeading: CGFloat = 0
+    @objc public var fullWidthGutterTrailing: CGFloat = 0
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationLayoutInfo.swift]: Tweaks the gutters, etc.

File: SignalServiceKit/src/Messages/Interactions/OWSDisappearingConfigurationUpdateInfoMessage.h [hunk 12]
```diff
@@ -20,6 +20,8 @@ NS_ASSUME_NONNULL_BEGIN
               createdByRemoteName:(nullable NSString *)remoteName
            createdInExistingGroup:(BOOL)createdInExistingGroup;
 
+- (NSString *)previewText;
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Messages/Interactions/OWSDisappearingConfigurationUpdateInfoMessage.h]: In some cases we can skip a transaction.

File: Signal/src/ViewControllers/MessageDetailViewController.swift [hunk 13]
```diff
@@ -83,6 +87,22 @@ class MessageDetailViewController: OWSViewController, MediaGalleryDataSourceDele
             object: OWSPrimaryStorage.shared().dbNotificationObject)
     }
 
+    override public func viewWillLayoutSubviews() {
+        Logger.debug("\(self.logTag) in \(#function)")
+
+        super.viewWillLayoutSubviews()
+
+        self.conversationLayoutInfo.viewWidth = self.view.width()
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/MessageDetailViewController.swift]: Most of the time the view size isn't changing. Would `viewWillTransition:toSize:` be a better candidate?

File: Signal/src/ViewControllers/ConversationView/ConversationViewLayout.m [hunk 14]
```diff
@@ -94,60 +94,42 @@ - (void)prepareLayout
...
-                break;
-            }
-            case ConversationViewLayoutAlignment_FullWidth:
-                itemFrame = CGRectMake(0, y, viewWidth, layoutSize.height);
-                break;
-            case ConversationViewLayoutAlignment_Center:
-                itemFrame = CGRectMake(
-                    hInset + round((viewWidth - layoutSize.width) * 0.5f), y, layoutSize.width, layoutSize.height);
-                break;
+        if (lastLayoutItem) {
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewLayout.m]: nit: `last` is ambiguous. I think you mean `previous`?

File: Signal/src/ViewControllers/DebugUI/DebugUIMessages.m [hunk 15]
```diff
@@ -1980,7 +1983,10 @@ + (DebugUIMessagesAction *)fakeRandomTextAction:(TSThread *)thread
                 OWSAssert(messageToQuote);
                 DDLogVerbose(@"%@ %@", self.logTag, label);
                 [DDLog flushLog];
-                ConversationViewItem *viewItem = [[ConversationViewItem alloc] initWithInteraction:messageToQuote isGroupThread:thread.isGroupThread transaction:transaction];
+                ConversationViewItem *viewItem = [[ConversationViewItem alloc] initWithInteraction:messageToQuote
+                                                                                     isGroupThread:thread.isGroupThread
+                                                                                       transaction:transaction
+                                                                                        layoutInfo:layoutInfo];
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/DebugUI/DebugUIMessages.m]: I'm somewhat resistant to caching the layoutInfo on the ConversationViewItem. As you can see here, we've perviously used them in multiple contexts, often outside of the ConversationViewController.

Also, it seems kind of awkward in the MessageDetailViewController.

If possible, it seems like the ConversationViewLayout or the ConversationViewController would be a better place to maintain a cache of any layout attributes.

File: Signal/src/ViewControllers/MessageDetailViewController.swift [hunk 16]
```diff
@@ -589,9 +605,7 @@ class MessageDetailViewController: OWSViewController, MediaGalleryDataSourceDele
             return
         }
 
-        messageBubbleView.contentWidth = contentWidth()
-
-        let messageBubbleSize = messageBubbleView.size(forContentWidth: contentWidth())
+        let messageBubbleSize = messageBubbleView.measureSize()
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/MessageDetailViewController.swift]: FWIW the old interface where messageBubbleView doesn't maintain state about it's layout context made more sense to me.  It seems like the VC should be directing the views, not the other way around.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 17]
```diff
@@ -252,6 +260,14 @@ - (nullable ConversationViewCell *)measurementCell
     return measurementCell;
 }
 
+- (CGFloat)vSpacingWithLastLayoutItem:(id<ConversationViewLayoutItem>)lastLayoutItem
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: s/Last/Previous ?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m [hunk 18]
```diff
@@ -62,19 +66,48 @@ - (void)commontInit
...
+    UIStackView *buttonStackView = [[UIStackView alloc] initWithArrangedSubviews:@[
+        self.addToContactsButton,
+        self.addToProfileWhitelistButton,
+        self.blockButton,
+    ]];
+    buttonStackView.axis = UILayoutConstraintAxisVertical;
+    buttonStackView.spacing = self.vSpacing;
+    // Ensure all of the buttons have the same width.
+    buttonStackView.alignment = UIStackViewAlignmentFill;
+    buttonStackView.layoutMargins = UIEdgeInsetsZero;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: For stack views, unless you're setting `isLayoutMarginsRelativeArrangement` I don't think layoutMargins matter.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m [hunk 19]
```diff
@@ -126,67 +159,36 @@ - (CGFloat)buttonVPadding
...
-    };
-
-    layoutButton(self.addToContactsButton, interaction.hasAddToContactsOffer);
-    layoutButton(self.addToProfileWhitelistButton, interaction.hasAddToProfileWhitelistOffer);
-    layoutButton(self.blockButton, interaction.hasBlockOffer);
+    return (self.buttonVPadding + CGSizeCeil([self.addToContactsButton sizeThatFits:CGSizeZero]).height);
 }
 
-- (CGSize)cellSizeForViewWidth:(int)viewWidth contentWidth:(int)contentWidth
+- (CGSize)cellSize
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: What is the motivation for removing the parameter and instead keeping layout info statefully on this model?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 20]
```diff
@@ -926,28 +926,32 @@ - (CGSize)bodyMediaSizeForContentWidth:(int)contentWidth
             }
 
             return CGSizeRound(CGSizeMake(mediaWidth, mediaHeight));
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: why this break after a return? Did you intend to leave this return?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 21]
```diff
@@ -51,11 +49,13 @@ - (void)commontInit
     // Ensure only called once.
     OWSAssert(!self.messageBubbleView);
 
-    _viewConstraints = [NSMutableArray new];
-
+    self.preservesSuperviewLayoutMargins = NO;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Isn't this default? Why this change?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 22]
```diff
@@ -256,11 +234,9 @@ - (void)updateDateHeader
         self.dateHeaderLabel.hidden = NO;
 
         [self.viewConstraints addObjectsFromArray:@[
-            // Date headers should be visually centered within the conversation view,
-            // so they need to extend outside the cell's boundaries.
-            [self.dateHeaderLabel autoSetDimension:ALDimensionWidth toSize:self.contentWidth],
-            (self.isIncoming ? [self.dateHeaderLabel autoPinEdgeToSuperviewEdge:ALEdgeLeading]
-                             : [self.dateHeaderLabel autoPinEdgeToSuperviewEdge:ALEdgeTrailing]),
+            // TODO: Are data headers symmetric or are they asymmetric? gutters are asymmetric?
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: re: this TODO

My understanding is that dates should be horizontally centered between the screen edges (not the the gutters).

Let's try to nip these little questions in the bud by asking the authority. Figuring out the right thing before opening the PR means less throw-away code for me to review. (and then re-review once you get the answer).

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 23]
```diff
@@ -432,10 +408,6 @@ - (CGSize)cellSizeForViewWidth:(int)viewWidth contentWidth:(int)contentWidth
         cellSize.height += self.footerHeight;
     }
 
-    if (self.shouldHaveFailedSendBadge) {
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Just making sure - this is intentionally going away with no replacement for now?

File: Signal/src/ViewControllers/ConversationView/ConversationLayoutInfo.swift [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+            SwiftAssertIsOnMainThread(#function)
+
+            updateProperties()
+        }
+    }
+
+    @objc public let contentMarginTop: CGFloat = 10
+    @objc public let contentMarginBottom: CGFloat = 10
+
+    @objc public var gutterLeading: CGFloat = 0
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationLayoutInfo.swift]: Are we really mutating instances of ConversationLayoutInfo? Or can these all be readonly?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m (hunk 3 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: We want to layout with regard to the cell _edges_, not it's margins.  Our design uses non-standard margins.  This is the default value and redundant with our setting the layout margins to zero as we do here, so I'll remove them - the intent was to make our intention clear (use non-standard layout with respect to edges) and future-proof us against future changes.

File: SignalServiceKit/src/Messages/Interactions/OWSDisappearingConfigurationUpdateInfoMessage.h (hunk 12 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/Interactions/OWSDisappearingConfigurationUpdateInfoMessage.h]: > It's true that sometimes we can skip using the transaction but, since we do need it sometimes we need to support having a transaction available to configure cells anyway, right?

We never need them for call/info/error messages which are rendered by "system message" cells, so I was trying to simply those cells.  But I've removed this since you don't like it.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSBubbleView.m [hunk 25]
```diff
@@ -165,64 +128,14 @@ - (void)updateLayers
 
 - (UIBezierPath *)maskPath
 {
-    return [self.class maskPathForSize:self.bounds.size
-                            isOutgoing:self.isOutgoing
-                              hideTail:self.hideTail
-                           isTruncated:self.isTruncated
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSBubbleView.m]: This was only used in `OWSBubbleView`'s mask generation logic - which didn't even consult that parameter, so I guess it vestigial.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m [hunk 26]
```diff
@@ -38,16 +41,17 @@ - (void)commontInit
 {
     OWSAssert(!self.titleLabel);
 
-    //    [self setTranslatesAutoresizingMaskIntoConstraints:NO];
+    self.preservesSuperviewLayoutMargins = NO;
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: See above - I was just trying to leave a reminder in these cells that - unlike our table view cells - they use non-standard margins.  As noted above, I've removed these.

File: SignalServiceKit/src/Messages/Interactions/OWSDisappearingConfigurationUpdateInfoMessage.h (hunk 12 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/Interactions/OWSDisappearingConfigurationUpdateInfoMessage.h]: Oh wait, I just remembered another reason why I did it - to avoid plumbing through a transaction in the measurement methods (e.g. `cellSize`) which also needs to do this kind of configuration.

File: Signal/src/ViewControllers/MessageDetailViewController.swift (hunk 13 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/MessageDetailViewController.swift]: Sure, that should be equivalent.

File: Signal/src/ViewControllers/ConversationView/ConversationViewLayout.m (hunk 14 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewLayout.m]: That's better, thanks.

File: Signal/src/ViewControllers/DebugUI/DebugUIMessages.m (hunk 15 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/DebugUI/DebugUIMessages.m]: * As an aside, I'm thinking of renaming `ConversationLayoutInfo` to `ConversationAppearance` or `ConversationStyle`, since I think it'll better capture the intention. 
* Conversation cell layout and styling (e.g. colors) is contextual (to the thread, the containing view size, etc.) so we definitely need either this kind of class or possibly a protocol/delegate pattern.
* `ConversationViewController` is a massive file - 5k+ lines.  I'm hoping that we can slowly decompose it into smaller components with well-defined responsibilities.  And I'm trying to resist it growing any larger.  The "conversation style" info seems like an opportunity to gather/bundle some related code.
* `ConversationViewLayout` is a layout for Conversation view's collection view - none of its logic or state is relevant to message details view.  Even if we did move these properties and logic there, we'd need to do just as much work to keep it's "view width" state updated.
* There are other alternatives that would make using this "appearance/style" class a little easier, but they would require KVO of the root view's frame & bounds, and the docs discourage that - and the hacks to do it are awkward and not future-proof.

Want to discuss when you come online?

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m (hunk 17 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: Done.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m (hunk 18 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: Removed.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 20 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: Fixed, thanks.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m (hunk 21 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Removed, see above.

File: Signal/src/ViewControllers/ConversationView/ConversationLayoutInfo.swift (hunk 24 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationLayoutInfo.swift]: * Some of the future properties (added in later PRs) are mutable, e.g. to reflect changes to dynamic type size.
* I'm trying to build this logic to make future work to support orientation changes easy, where these views will automatically update to reflect changes to the root view size.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m (hunk 23 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: It's on my list of questions to review with Myles.  I'm trying to batch my questions to him (I already have ~30 from just a day's work).

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m (hunk 22 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: It's on my list of questions to review with Myles.  I'm trying to batch my questions to him (I already have ~30 from just a day's work).

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m (hunk 19 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSContactOffersCell.m]: See above - there's now a ton of "constants" and "semi-constants" the are calculated from each other.  I'm trying to avoid: 

* Passing them all around separately
* Calculating them all for each cell.

We could maintain an instance of "conversation layout info/style/appearance" and pass it into the cell's "configure" and "measurement" methods?  Would that feel better to you?  If it's easy, I'd be happy to do it.

**Comment by charlesmchen-signal** [comment]
PTAL @michaelkirk

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 27]
```diff
@@ -963,31 +967,28 @@ - (CGSize)quotedMessageSizeForContentWidth:(int)contentWidth includeMargins:(BOO
...
     cellSize.width = MAX(cellSize.width, quotedMessageSize.width);
     cellSize.height += quotedMessageSize.height;
 
-    CGSize mediaContentSize = [self bodyMediaSizeForContentWidth:contentWidth];
+    CGSize mediaContentSize = [self bodyMediaSize];
     cellSize.width = MAX(cellSize.width, mediaContentSize.width);
     cellSize.height += mediaContentSize.height;
 
-    CGSize textContentSize = [self bodyTextSizeForContentWidth:contentWidth includeMargins:YES];
+    CGSize textContentSize = [self bodyTextSize:YES];
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: nit: Can you label the parameter in the message signature e.g. `bodyTextSizeWithIncludeMargins:YES`?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 28]
```diff
@@ -411,16 +384,17 @@ - (UIFont *)dateHeaderTimeFont
...
+    OWSAssert(self.layoutInfo.viewWidth > 0);
     OWSAssert(self.viewItem);
     OWSAssert([self.viewItem.interaction isKindOfClass:[TSMessage class]]);
     OWSAssert(self.messageBubbleView);
 
     self.messageBubbleView.viewItem = self.viewItem;
-    self.messageBubbleView.contentWidth = self.contentWidth;
     self.messageBubbleView.cellMediaCache = self.delegate.cellMediaCache;
-    CGSize messageBubbleSize = [self.messageBubbleView sizeForContentWidth:contentWidth];
+    CGSize messageBubbleSize = [self.messageBubbleView measureSize];
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: I was surprised that `measureSize` didn't ever need a transaction to compute it's sizing, but it looks like everything it uses is already cached on the viewModel. 👍

File: Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m [hunk 29]
```diff
@@ -240,48 +264,31 @@ - (CGFloat)iconSize
...
+    CGSize result = CGSizeMake(self.layoutInfo.viewWidth, 0);
 
-    // FIXME pass in transaction from the uiDBConnection.
-    [[TSYapDatabaseObject dbReadConnection] readWithBlock:^(YapDatabaseReadTransaction *_Nonnull transaction) {
-        [self applyTitleForInteraction:interaction label:self.titleLabel transaction:transaction];
-    }];
-
-    CGFloat maxTitleWidth = (contentWidth - ([self hMargin] * 2.f + [self hSpacing] + [self iconSize]));
-    CGSize titleSize = [self.titleLabel sizeThatFits:CGSizeMake(maxTitleWidth, CGFLOAT_MAX)];
+    [self applyTitleForInteraction:interaction label:self.titleLabel transaction:transaction];
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSSystemMessageCell.m]: 🎉

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 27 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: Done.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m (hunk 28 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: Yup, that's where I'd like to get for all of this.
